### PR TITLE
[lldb][riscv] Fix setting breakpoint for undecoded instruction 

### PR DESCRIPTION
**Pull request by ita-sc**
Hi

This patch adds an interface GetLastInstrSize to get information about the size of last tried to be decoded instruction and uses it to set software breakpoint if the memory can be decoded as instruction.

RISC-V architecture instruction format specifies the length of instruction in first bits, so we can set a breakpoint for these cases. This is needed as RISCV have a lot of extensions, that are not suppored by `EmulateInstructionRISCV`.

### CODE REVIEW

File: lldb/source/Plugins/Instruction/RISCV/EmulateInstructionRISCV.h
Comment (jasonmolenda) [inline lldb/source/Plugins/Instruction/RISCV/EmulateInstructionRISCV.h:103]: I might call this `Last decoded instruction size`.

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (jasonmolenda) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:151]: Shouldn't this block now be `if (emulator_up->ReadInstruction())` now?  We're going to get the size of the last decoded instruction here.

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (jasonmolenda) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:155]: We've defined the new GetLastInstrSize() method for the RISCV EmulateInstruction plugin, but others like AArch64 won't have that, so this will error out on them, won't it?  What we really want to express is "if `arch.GetTriple().isRISCV()` and we couldn't decode the length of the last instruction, then error out" isn't it?

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (jasonmolenda) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:164]: We've decoded the length of the instruction at `pc` at this point, and them to get `next_pc`.  Then we pass `next_pc` to this method which has a hardcoded size of 4 for RISCV.  It's only a hint that is sent to lldb-server as it tries to step over the instruction.  With armv7/aarch32 we had to get arm/thumb breakpoint instructions correct because an arm breakpoint wasn't valid when the processor was in thumb mode (iirc) but RISC-V doesn't have a processor mode like that iiuc.  So maybe it's fine to have `SetSoftwareBreakPointOnPC` hardcoding 4 for the next RISCV breakpoint.

**Review by jasonmolenda — COMMENTED** [review]
I read this patch while trying to keep the RVC instruction set in mind because it's an interesting case to consider.  Maybe the test file could have a compressed instruction?  The one instruction there now ends in 0b11 so I don't think it will decode that way.

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (jasonmolenda) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:151]: Ah wait, I see.  This method is trying to decode where the next instruction will go, with branches and jumps decoded, so we can put a breakpoint there.  And you're handling the case where we can't decode the current instruction (I now understand why you used that in your test case).  It seems harmless to call GetLastInstrSize() if the instruction that couldn't be decoded, and add the length of the instruction to pc.  We can assume the emulation engine will emulate all branching instructions.  I could imagine the RISCV emulation plugin didn't have decoding for an instruction that doesn't branch, it could fail, but we can still decode the size of that unknown instruction successfully, and assume that it does not branch.

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:155]: Well, in this MR this is a common code, and every other architecture now not implemented `GetLastInstrSize`, so it will fail with the same error, as it was before. For RISC-V we will generate the same error if we can not get the size of instruction (if we can not read memory or if an instruction is bigger than expected)

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:164]: Actually, I do not understand fully what you mean. For RISCV software breakpoint size is 4 or 2 (currently LLDB hardcodes only 4). We save memory and restore it later, so it seems we will have no problems here for different instruction sizes on memory regions where we write software breakpoint.

File: lldb/source/Plugins/Instruction/RISCV/EmulateInstructionRISCV.h
Comment (ita-sc) [reply lldb/source/Plugins/Instruction/RISCV/EmulateInstructionRISCV.h]: Well, I tried to emphasize that even if we failed to fully decode instruction, we are still able to fill this information. I'll change to 
```
Last decoded instruction size estimate.
```

**Comment by ita-sc** [comment]
> I read this patch while trying to keep the RVC instruction set in mind because it's an interesting case to consider. Maybe the test file could have a compressed instruction? The one instruction there now ends in 0b11 so I don't think it will decode that way.

The test intentionally has an invalid instruction. I'll add one more test with compressed illegal instruction.
In current test it seems it is decoded as some 4-size instruction, as expected: https://godbolt.org/z/bMPf7faWW
```
addi	s0,sp,16
 .4byte	0xc58573
```

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (labath) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:97]: Doesn't this return the size of the *break*point (i.e. it should be called `GetSoftwareBreakpointSize` ?

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (labath) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:113]: We usually spell "break point" as a single word (i.e., `Breakpoint`).

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (labath) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:124]: https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (labath) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:153]: https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

The rule is pretty subjective, but I think the fact you felt the need to add "opt" to the name shows the code is not completely understandable without it, and I'd say that spelling out the name is the more conventional way to express that. Also, (for better or worse) most of the lldb code (the surrounding code included) uses snake_case for variable name, so it's better to stick to that.

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp]: Fixed

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp]: Fixed

File: lldb/test/API/riscv/break-undecoded/TestBreakpointIlligal.py
Comment (ita-sc) [reply lldb/test/API/riscv/break-undecoded/TestBreakpointIlligal.py]: Fixed

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp]: Changed to `instr_size`

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp]: Fixed, changed to `GetSoftwareBreakpointSize`

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:103]: `else` after return

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (JDevlieghere) [inline lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:113]: Small nit, but in recent years we've been preferring using `llvm::Error` over `lldb::Status` where possible. The benefit of `Error` is that it must be checked and it's trivial to convert between the two. It would be nice if this function would return an `llvm::Error`.

**Review by JDevlieghere — COMMENTED** [review]
A few small nits. I defer to Pavel and Jason for the functional changes.

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp]: Fixed

File: lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp
Comment (ita-sc) [reply lldb/source/Plugins/Process/Utility/NativeProcessSoftwareSingleStep.cpp:112]: I would like to keep `Status` in this MR, as from my point of view, this MR will do more that it should -- it will add a new functionality and change the way we work with errors here.

Moreover, changing this will result in unnecessary casts from `Status` to `Error` (`process.SetBreakpoint`) and from `Error` to `Status`(`SetupSoftwareSingleStepping`).

Note: I've checked that we have `Status::ToError` and `Status(llvm::Error error)` ctor, but I'm not sure these functions will suffice (AFAIU `m_code` will be saved only if `m_type == ErrorType::eErrorTypePOSIX`).

**Comment by ita-sc** [comment]
Gentle ping

**Review by labath — APPROVED** [review]
Thanks for pinging this. Sorry it dropped off my radar.

**Comment by ita-sc** [comment]
Thanks. Could you please merge this? (I do not write access yet)

**Pull request merged**
